### ISSUE TITLE
Cannot compile to target windows_i386

### ISSUE THREAD

**Issue by Apis035** [odin-lang/Odin#3091]
## Context

        Odin: dev-2024-01:5d94887e
        OS:   Windows 10 Professional (version: 22H2), build 19045.3693
        CPU:  Intel(R) Core(TM) i5-6300U CPU @ 2.40GHz
        RAM:  7591 MiB

## Expected Behavior

Odin successfully compile a Windows 32-bit executable.

## Current Behavior

Linker complains about unresolved external symbol.

## Steps to Reproduce

```
E:\test>type test.odin
package test
import "core:fmt"
main :: proc() { fmt.println("Hello Odin!") }

E:\test>odin build . -target:windows_i386
libcmt.lib(exe_main.obj) : error LNK2019: unresolved external symbol _main referenced in function "int __cdecl __scrt_common_main_seh(void)" (?__scrt_common_main_seh@@YAHXZ)
E:\test\test.exe : fatal error LNK1120: 1 unresolved externals
```

**Comment by flysand7** [comment]
Seems like there is a bug in Windows entry point.

```odin
when ODIN_ARCH == .i386 || ODIN_NO_CRT {
	@(link_name="mainCRTStartup", linkage="strong", require)
	mainCRTStartup :: proc "stdcall" () -> i32 {
		context = default_context()
		#force_no_inline _startup_runtime()
		intrinsics.__entry_point()
		#force_no_inline _cleanup_runtime()
		return 0
	}
} else {
	@(link_name="main", linkage="strong", require)
	main :: proc "c" (argc: i32, argv: [^]cstring) -> i32 {
		args__ = argv[:argc]
		context = default_context()
		#force_no_inline _startup_runtime()
		intrinsics.__entry_point()
		#force_no_inline _cleanup_runtime()
		return 0
	}
}
```

Seems like by default Odin doesn't define `main` CRT entry point for 32-bit windows. I'm not sure why i386 was put in here, maybe someone else can explain why it is like that?

**Comment by flysand7** [comment]
@Apis035 does it work if you manually go into that file and remove `ODIN_ARCH == .i386` in the `when` condition in `core/runtime/entry_windows.odin`?

**Comment by Apis035** [comment]
yes, it does work when `ODIN_ARCH == .i386` is removed

**Comment by Apis035** [comment]
compilation also works without modification to that file when compiling with `-no-crt`

**Comment by zhylmzr** [comment]
After removing the `ODIN_ARCH == .i386`  from `entry_windows.odin`, and compiling with `-no-thread-local` works fine.

**Comment by flysand7** [comment]
> Tried with -no-crt as suggested, but the linker spits that __chkstk is unresolved.

Yes, `__chkstk` is a function that is provided by the *libc runtime* and is used by compiler to commit the virtual memory of the stack, when large stack allocations (about `>= page_size`) happen.

For security reasons the stack memory 2 page sizes below the stack pointer is not accessible, and is allocated further only when the **next page of the stack** is used. The `__chkstk` function iterates from the stack pointer down and performs a dummy-write to every page of the allocated addresses to ensure all of it is committed into memory.

For `-no-crt` compilation, that would mean that the Odin's runtime library would need to provide this function. Since it involves getting the value of the stack pointer that would also mean it probably has to be written in assembly.

I only have the version of this function for `x86-64`, which I had obtained by disassembling a running executable and cleaning it up:

https://github.com/flysand7/ciabatta/blob/27ee227f8e40ead63856f5e7a59d37660ee26620/arch/x64_x86-64/chkstk.asm#L2-L26

@gingerBill, @laytan I'm surprised that the `__chkstk` issue doesn't pop up for x86-64, is there something on the llvm/linker front that is disabled? Because I seriously doubt no one has ever allocated an array bigger than 8096 bytes in size.

---

Actually, upon reading my above response I realised, that I don't actually know *who* is responsible for generating the call to the `__chkstk` intrinsic. Is it our compiler? Is it LLVM?

**Comment by laytan** [comment]
@flysand7 

> Actually, upon reading my above response I realised, that I don't actually know who is responsible for generating the call to the __chkstk intrinsic. Is it our compiler? Is it LLVM?

LLVM

> I'm surprised that the __chkstk issue doesn't pop up for x86-64, is there something on the llvm/linker front that is disabled?

It is not an issue because YOU contributed a __chkstk for that arch right (base/procs_windows_amd64.asm)? Or am I missing some context about the issue here.

**Comment by flysand7** [comment]
> YOU contributed a __chkstk for that arch

Honestly I forgot that I did that lmao

**Comment by Feoramund** [comment]
So does i386 Windows just need a `__chkstk`?

**Comment by Kelimion** [comment]
I get a different error:
```
> odin build . -vet -vet-tabs -strict-style -vet-style -warnings-as-errors -disallow-do -out:scratch.exe -target:windows_i386

libcmt.lib(exe_main.obj) : error LNK2019: unresolved external symbol _main referenced in function "int __cdecl __scrt_common_main_seh(void)" (?__scrt_common_main_seh@@YAHXZ)
W:\Scratch\scratch.exe : fatal error LNK1120: 1 unresolved externals
```

**Comment by Kelimion** [comment]
If I comment out `ODIN_ARCH == .i386`, then:
```
odin build . -vet -vet-tabs -strict-style -vet-style -warnings-as-errors -disallow-do -out:scratch386.exe -target:windows_i386
scratch386.exe
echo %ERRORLEVEL%
-1073741819
```
```
W:\Scratch>dumpbin scratch386.exe /HEADERS
Microsoft (R) COFF/PE Dumper Version 14.43.34810.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file scratch386.exe

PE signature found

File Type: EXECUTABLE IMAGE

FILE HEADER VALUES
             14C machine (x86)
               5 number of sections
        68482061 time date stamp Tue Jun 10 14:09:05 2025
               0 file pointer to symbol table
               0 number of symbols
              E0 size of optional header
             102 characteristics
                   Executable
                   32 bit word machine

OPTIONAL HEADER VALUES
             10B magic # (PE32)
           14.43 linker version
           54A00 size of code
           2E200 size of initialized data
               0 size of uninitialized data
           48A8C entry point (00448A8C)
            1000 base of code
           56000 base of data
          400000 image base (00400000 to 00485FFF)
            1000 section alignment
             200 file alignment
            6.00 operating system version
            0.00 image version
            6.00 subsystem version
               0 Win32 version
           86000 size of image
             400 size of headers
               0 checksum
               3 subsystem (Windows CUI)
```
Does look like a 32-bit PE file.

**Comment by Apis035** [comment]
I think this can be closed, compiling with `odin build . -target:windows_i386` does not error anymore.
